webpack/localization-plugin/src/LocalizationPlugin.ts

```diff
@@ -0,0 +1,732 @@
+   * If {@link IDefaultLocaleOptions.usePassthroughLocale} is set, use this name for the passthrough locale.
+   * Defaults to "passthrough"
+   */
+  passthroughLocaleName?: string;
+}
+
+/**
+ * The options for localization.
+ *
+ * @public
+ */
+export interface ILocalizationPluginOptions {
+  localizedStrings: ILocales;
+  defaultLocale: IDefaultLocaleOptions;
+  serveLocale: IDefaultLocaleOptions;
+  filesToIgnore?: string[];
+  localizationStatsDropPath?: string;
+  localizationStatsCallback?: (stats: ILocalizationStats) => void;
+}
+
+/**
+ * @internal
+ */
+export interface IStringPlaceholder {
+  value: string;
+  suffix: string;
+}
+
+const PLUGIN_NAME: string = 'localization';
+const LOCALE_FILENAME_PLACEHOLDER: string = '[locale]';
+const LOCALE_FILENAME_PLACEHOLDER_REGEX: RegExp = new RegExp(lodash.escapeRegExp(LOCALE_FILENAME_PLACEHOLDER), 'g');
+const STRING_PLACEHOLDER_PREFIX: string = '-LOCALIZED-STRING-f12dy0i7-n4bo-dqwj-39gf-sasqehjmihz9';
+
+interface IProcessAssetResult {
+  filename: string;
+  asset: IAsset;
+}
+
+interface IAsset {
+  size(): number;
+  source(): string;
+}
+
+interface IExtendedMainTemplate {
+  hooks: {
+    localVars: Tapable.SyncHook<string, Webpack.compilation.Chunk, string>;
+  };
+}
+
+interface ISingleLocaleConfigOptions {
+  localeName: string;
+  resolvedStrings: Map<string, Map<string, string>>;
+  passthroughLocale: boolean;
+}
+
+/**
+ * @public
+ */
+export interface ILocalizationStatsEntrypoint {
+  localizedAssets: { [locale: string]: string };
+}
+
+/**
+ * @public
+ */
+export interface ILocalizationStatsChunkGroup {
+  localizedAssets: { [locale: string]: string };
+}
+
+/**
+ * @public
+ */
+export interface ILocalizationStats {
+  entrypoints: { [name: string]: ILocalizationStatsEntrypoint };
+  namedChunkGroups: { [name: string]: ILocalizationStatsChunkGroup };
+}
+
+/**
+ * This plugin facilitates localization in webpack.
+ *
+ * @public
+ */
+export class LocalizationPlugin implements Webpack.Plugin {
+  /**
+   * @internal
+   */
+  public stringKeys: Map<string, IStringPlaceholder>;
+
+  private _options: ILocalizationPluginOptions;
+  private _locJsonFiles: Set<string>;
+  private _locJsonFilesToIgnore: Set<string>;
+  private _stringPlaceholderCounter: number;
+  private _stringPlaceholderMap: Map<string, { [locale: string]: string }>;
+  private _passthroughStringsMap: Map<string, string>;
+  private _locales: Set<string>;
+  private _localeNamePlaceholder: IStringPlaceholder;
+  private _defaultLocale: string;
+  private _serveLocale: string;
+  private _usePassthroughForServe: boolean | undefined;
+  /**
+   * The outermost map's keys are the locale names.
+   * The middle map's keys are the resolved, uppercased file names.
+   * The innermost map's keys are the string identifiers and its values are the string values.
+   */
+  private _resolvedLocalizedStrings: Map<string, Map<string, Map<string, string>>>;
+
+  public constructor(options: ILocalizationPluginOptions) {
```

Comment:
How will this work in our projects? Will we need to include large `ILocalizationPluginOptions` settings in the webpack config, or will this be generated somehow?

Also, if I modify a `.loc.json` file, will I need to amend the config and restart webpack?

Reply:
For the time being, yes. Right now this is oversimplified, but I want that to be improved soon.